docs/architecture/adr-014-versioned-namespaces.md

```diff
@@ -0,0 +1,192 @@
+
+## Implementation Details
+
+Regardless of the option chosen, end-users must explicitly specify the namespace ID version (and/or) share version they intend on using in their PFB transaction.
+
+Since the NMT needs to be aware of the version byte, Option A is equivalent to increasing the namespace ID size to `9` bytes and then constraining the namespaces available for use to namespaces that have a leading `0` byte. In other words:
+
+- MinReservedNamespace: []byte{0, 0, 0, 0, 0, 0, 0, 0, 1}
+- MaxReservedNamespace: []byte{0, 0, 0, 0, 0, 0, 0, 0, 255}
+- MinBlobNamespace: []byte{0, 0, 0, 0, 0, 0, 0, 1, 0}
+- MaxBlobNamespace: []byte{0, 255, 255, 255, 255, 255, 255, 255, 255}
+
+When a user creates a PFB, concatenate the namespace version with the namespace ID to derive the namespace that is pushed to the NMT. Option C is similar to Option A, if we constrain the available share versions at mainnet to `0`.
+
+## Open Questions
+
+1. Is the scenario where we modify non-interactive default rules for a subset of namespaces contrived? When would we want to support a data square with two different non-interactive default rules? IMO roll-ups shouldn’t make assumptions about the padding in-between blobs so we shouldn’t need to support backwards compatibility with respect to padding changes.
+2. When do we expect to increment the existing share version?
+    1. Option 1: when there are changes to the universal share prefix
+    2. Option 2: when there are changes to any part of the remaining data in a share
+3. When do we expect to increment the namespace version?
+    1. During a backwards incompatable non-interactive default rule change
+    2. If we change the format of a padding share (e.g. a namespace padding share) instead of `0` bytes, pad with something else like. We may need to preserve backwards compatibility for padding shares that use old namespaces. Note this scenario likely implies a namespace version and share version increase.
+    3. Change the format of PFB tx serialization. This scenario likely implies duplicating the PFB txs in a data square, one with the old namespace version and one with the new namespace version.
+4. Inspired by [type-length-value](https://en.wikipedia.org/wiki/Type%E2%80%93length%E2%80%93value), should we consider prefixing optional fields (sequence length and reserved bytes) with a type and a length? This would enable us to modify those fields without introducing new share versions.
+5. [Requires investigation] what changes need to be made to NMT in order to support namespaces of a different length (e.g. 16 bytes)?
+6. [Requires investigation] what changes need to be made to support variable length namespace IDs? In other words, roll-ups may use a namespace ID of 8 - 32 bytes. Are variable length namespace IDs something we are considering supporting?
```

Comment:
is this something that we are still wanting to investigate? I don't think we were explicit about it, but my understanding was that it would be too messy, although possible

Reply:
I'm going to investigate 

> 5. [Requires investigation] what changes need to be made to NMT in order to support namespaces of a different length (e.g. 16 bytes)?

in another branch / PR with the assumption that we increase namespace ID size to 16 bytes (or higher). Will punt investigating variable length namespace IDs for now

Reply:
@evan-forbes is it safe to resolve this convo?